### PR TITLE
Be less tolerant of broken installs.

### PR DESCRIPTION
**Pull request by anntzer**
If no user-defined matplotlibrc is present, require it to exists in
mpl-data.  (We already throw when mpl-data cannot be found -- which is
also a sign of broken install -- so it's not much worse to require
matplotlibrc to exist as well.)

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

**Comment by ImportanceOfBeingErnest** [comment]
Can you comment on why it would be necessary for *some* rc file to exist? It's usually a completely commented out file, so it serves no purpose, and does no harm if it's not found.

**Comment by anntzer** [comment]
Because the build script creates it (at https://github.com/matplotlib/matplotlib/blob/master/setup.py#L204).
We have tooling to ensure that a file exists, we shouldn't need additional tooling to cover the case where it doesn't.

**Comment by ImportanceOfBeingErnest** [comment]
I'd say a user-friendly software would try to repair itself if it finds itself in a currupted state. I mean it's not obvious from anywhere that this file needs to exist.

**Comment by anntzer** [comment]
You also really need to mess it up/do it on purpose to have an install where the file doesn't exist (it's like an install where axes/_base.py doesn't exist, sure you could have deleted it accidentally but other than that...).
(Alternatively I'd be happy to see a realistic case where the file would be missing.)

**Comment by anntzer** [comment]
We already look in $prefix/matplotlib/mpl-data/matplotlibrc where $prefix/matplotlib is the place matplotlib is actually installed.  And that file is created at install time, in fact it's the one that gets picked up when you haven't configured your own matplotlibrc.

**Comment by tacaswell** [comment]
> We have tooling to ensure that a file exists, we shouldn't need additional tooling to cover the case where it doesn't.

I am not sure we want to impose this on our down-stream packagers.  We know that various linux distributions significantly de-construct our installation.

attn @sandrotosi .

**Comment by QuLogic** [comment]
In fact, I ran into issues with the fallback when the rc file did not exist. We used to have an rc file that just specified the backend, but dropped that since mpl can pick one automatically now. Since we patch the system location to be in `/etc/matplotlibrc` but testing is done on a temporary build root, mpl could not find the defaults, and triggered the [fallback code that's removed here](https://github.com/matplotlib/matplotlib/pull/12246/files#diff-9ffb0d1db51a67ee4f37528e00715b3cL922). I did not investigate further, but that fallback code _does not work_.

Whatever happens when an rc file (even if the rc file just specifies a backend and nothing else) is loaded didn't happen with the fallback, and a lot of params were their unvalidated non-canonical forms that didn't work when actually used.

TLDR, we already patch `gen_candidates` to use the location we want, and it never triggered this fallback code before, so it doesn't matter if it's not there.

**Comment by tacaswell** [comment]
I am going to merge this as is, if we find it to be a problem, we should add a fallback of reading an empty file from a temp location.

**Comment by sandrotosi** [comment]
sorry for the late reply, but this is fine even for Debian - we force `/etc/matplotlib` into the paths to check and we make sure that file is present by installing it in the deb package
